File: crates/saber-provider/src/retry.rs
//! Basic exponential-backoff retry (M0). The full failover chain
//! (switching to a fallback provider/model) lands with M1's router; this
//! wrapper retries the same request on retryable failures.
//!
//! Because runtime errors surface as the stream's first event, the wrapper
//! peeks one event: a retryable `Error` head consumes an attempt, anything
//! else is passed through untouched.

use crate::{ChatRequest, Provider, ProviderEvent, ProviderStream, RetryKind};
use futures::future::BoxFuture;
use futures::stream::{StreamExt, once};
use std::time::Duration;

#[derive(Debug, Clone)]
pub struct RetryPolicy {
    pub max_attempts: u32,
    pub base_delay: Duration,
}

impl Default for RetryPolicy {
    fn default() -> Self {
        Self {
            max_attempts: 3,
            base_delay: Duration::from_millis(500),
        }
    }
}

/// Runs `provider.stream(request)` with backoff on retryable failures and
/// returns the first attempt that starts cleanly.
pub fn stream_with_retry(
    provider: std::sync::Arc<dyn Provider>,
    request: ChatRequest,
    policy: &RetryPolicy,
) -> BoxFuture<'static, ProviderStream> {
    let policy = policy.clone();
    Box::pin(async move {
        let mut attempt: u32 = 0;
        loop {
            attempt += 1;
            let stream = match provider.stream(request.clone()).await {
                Ok(stream) => stream,
                Err(e) => {
                    if attempt < policy.max_attempts {
                        backoff(attempt, &policy).await;
                        continue;
                    }
                    return fatal(format!("{e}"));
                }
            };
            let mut stream = stream;
            let Some(first) = stream.next().await else {
                return fatal("provider returned an empty stream".to_owned());
            };
            let retryable_head = matches!(
                &first,
                ProviderEvent::Error {
                    retryable: RetryKind::RateLimit
                        | RetryKind::Timeout
                        | RetryKind::ServerError
                        | RetryKind::Network,
                    ..
                }
            );
            if retryable_head && attempt < policy.max_attempts {
                backoff(attempt, &policy).await;
                continue;
            }
            return once(async move { first }).chain(stream).boxed();
        }
    })
}

async fn backoff(attempt: u32, policy: &RetryPolicy) {
    let shift = (attempt - 1).min(6);
    let delay = policy.base_delay.mul_f64(2f64.powi(shift as i32));
    tokio::time::sleep(delay).await;
}

fn fatal(message: String) -> ProviderStream {
    once(async move {
        ProviderEvent::Error {
            message,
            retryable: RetryKind::Fatal,
        }
    })
    .boxed()
}

File: packages/ai/src/retry.ts
import type { ChatRequest, Provider, ProviderEvent } from "./types.ts";

export interface RetryPolicy {
  maxAttempts: number;
  baseDelayMs: number;
}

export const defaultRetryPolicy: RetryPolicy = { maxAttempts: 3, baseDelayMs: 500 };

export async function* streamWithRetry(
  provider: Provider,
  request: ChatRequest,
  policy: RetryPolicy = defaultRetryPolicy,
): AsyncGenerator<ProviderEvent> {
  for (let attempt = 1; attempt <= policy.maxAttempts; attempt++) {
    let firstEvent: ProviderEvent | null = null;
    for await (const event of provider.stream(request)) {
      if (!firstEvent) {
        firstEvent = event;
        if (event.type === "error" && event.retryable !== "fatal" && attempt < policy.maxAttempts) {
          await sleep(policy.baseDelayMs * Math.pow(2, attempt - 1));
          break; // retry
        }
      }
      yield event;
      if (firstEvent === event && event.type === "error") return; // fatal, no retry
    }
    if (firstEvent && firstEvent.type !== "error") return; // success
    if (attempt === policy.maxAttempts && firstEvent) {
      yield firstEvent; // give up, surface the error
      return;
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
